Form based injection configuration (#185)
jenkinsci/gradle-plugin: README.adoc

== Gradle Enterprise integration

The plugin can be configured to inject the https://docs.gradle.com/enterprise/gradle-plugin/[Gradle Enterprise Gradle plugin] or the https://docs.gradle.com/enterprise/maven-extension/[Gradle Enterprise Maven extension] into any Gradle or Maven build that is executed on the Jenkins server or on any of its connected agents.

> To achieve this, the plugin installs various files on each agent depending on the configuration via the global options in *Manage Jenkins*, *Configure System*.

Reviewer: ```suggestion
To achieve this, the plugin installs various files on each agent depending on the configuration via the global options in the `Manage Jenkins/Configure System` section.
```

The same auto-injection behavior is available for the https://github.com/gradle/common-custom-user-data-gradle-plugin[Common Custom User Data Gradle plugin] and the https://github.com/gradle/common-custom-user-data-maven-extension[Common Custom User Data Maven extension].

> **Note**
> - The configuration applies to all builds on all connected agents matching the specified label criteria, or all in case no label criteria are defined.

=== Enable auto-injection

The auto-injection is split into three levels to have fine-grained control over the enablement of this feature.

==== Global

> To globally enable the auto-injection, click the `Enable auto-injection` checkbox in the `Gradle Enterprise integration` section of the global options in *Manage Jenkins*, *Configure System*.

Reviewer: ```suggestion
To globally enable the auto-injection, click the `Enable auto-injection` checkbox in the `Gradle Enterprise integration` section of the global options in the `Manage Jenkins/Configure System` section.
```

> Next set the URL of the Gradle Enterprise instance.

Reviewer: ```suggestion
Next set the URL of the Gradle Enterprise instance to which the build scans should be published to.
```

image::gradle-settings.png[Gradle settings,{thumbnail}]

Optionally set the desired version of the https://github.com/gradle/common-custom-user-data-gradle-plugin[Common Custom User Data Gradle plugin] to be used.

==== Maven

To enable the auto-injection for Maven builds, click the `Enable Gradle Enterprise Maven extension auto-injection` checkbox in the `Maven settings` section of the configuration form.

image::maven-settings.png[Maven settings,{thumbnail}]

Optionally click the `Enable Common Custom User Data Maven extension auto-injection` checkbox to enable the auto-injection of the https://github.com/gradle/common-custom-user-data-maven-extension[Common Custom User Data Maven Extension].

To see which versions are injected, refer to <<Compatibility>>.

> **Warning**
> - Maven injection only works if `MAVEN_OPTS` is not configured as a global environment variable.

=== Disable auto-injection

Disabling the auto-injection requires that all Gradle Enterprise resources are cleaned up from the agents.
To achieve this, the Gradle or Maven injections must be disabled individually (see the following sections).
This triggers a cleanup of the resources.

> If the auto-injection should be disabled globally, then unset the `Enable auto-injection` checkbox as well.

Reviewer: ```suggestion
If the auto-injection should be disabled globally, then uncheck the `Enable auto-injection` checkbox as well.
```

==== Gradle

To disable the auto-injection for Gradle builds, remove the `Gradle Enterprise Gradle plugin version` in the `Gradle settings` section of the configuration form.

==== Maven

> To disable the auto-injection for Maven builds, unset the `Enable Gradle Enterprise Maven extension auto-injection` checkbox in the `Maven settings` section of the configuration form.

Reviewer: ```suggestion
To disable the auto-injection for Maven builds, uncheck the `Enable Gradle Enterprise Maven extension auto-injection` checkbox in the `Maven settings` section of the configuration form.
```

=== Control auto-injection based on node labels

> Auto-injection can be enabled or disabled based on specific node labels.
> To enable auto-injection only on specific nodes add the desired labels to the `Gradle auto-injection enabled nodes` list for Gradle or `Maven auto-injection enabled nodes` for Maven.
> To disable auto-injection on specific nodes add the desired labels to the `Gradle auto-injection disabled nodes` list for gradle or `Maven auto-injection disabled nodes` for Maven.

Reviewer: ```suggestion
Auto-injection can be enabled or disabled based on specific node labels.
To enable auto-injection only on specific nodes, add the desired labels to the `Gradle auto-injection enabled nodes` list for Gradle or `Maven auto-injection enabled nodes` for Maven.
To disable auto-injection on specific nodes, add the desired labels to the `Gradle auto-injection disabled nodes` list for Gradle or `Maven auto-injection disabled nodes` for Maven.
```

The disabled labels list will take precedence over the enabled labels list.

> === Disabling/removing Gradle Jenkins plugin
> 
> To ensure that all Gradle Enterprise resources are cleaned up from the agents, before disabling/removing the plugin the auto-injection has to be manually disabled. Please, refer to the <<Disable auto-injection>> section for details.

Reviewer: ```suggestion
== Uninstalling Gradle Jenkins plugin

To ensure that all Gradle Enterprise resources are cleaned up from the agents, before uninstalling the plugin the auto-injection has to be manually disabled. Please, refer to the <<Disable auto-injection>> section for details.
```

Please also move that top level after the `Gradle Enterprise integration`. (See my change on heading level too)

Refer to the https://docs.gradle.com/enterprise/gradle-plugin/#via_environment_variable[Gradle Enterprise Gradle plugin manual] and the https://docs.gradle.com/enterprise/maven-extension/#via_environment_variable[Gradle Enterprise Maven Extension manual] for more details about this configuration.

==== Gradle specific configuration options

`Gradle Enterprise Gradle plugin version` +
Enables auto-injection for Gradle builds and defines which version of the https://docs.gradle.com/enterprise/gradle-plugin/[Gradle Enterprise Gradle plugin] to use.

`Common Custom User Data Gradle plugin version` +
Defines which version of the https://github.com/gradle/common-custom-user-data-gradle-plugin[Common Custom User Data Gradle plugin] to use.

`Gradle plugin repository url` +
The URL of the repository to use to resolve the Gradle Enterprise Gradle plugin and the Common Custom User Data Gradle plugin.
This is required if the Jenkins agents are not able to access the https://plugins.gradle.org[Gradle Plugin Portal].

`Gradle auto-injection enabled nodes` +

> A list of node labels on which the Gradle Enterprise Gradle plugin injection should be enabled.

Reviewer: ```suggestion
A list of node labels on which the Gradle Enterprise Gradle plugin or Common Custom User Data Gradle plugin injection should be enabled.
```

Please adapt this on all 4 sections about node labels (Gradle and Maven)

Author: Done

By default, all nodes are enabled.

`Gradle auto-injection disabled nodes` +
A list of node labels on which the Gradle Enterprise Gradle plugin injection should be disabled.
By default, all nodes are enabled.

==== Maven specific configuration options

`Enable Gradle Enterprise Maven extension auto-injection` +
Enables auto-injection for Maven builds and injects https://docs.gradle.com/enterprise/maven-extension[Gradle Enterprise Maven extension].

`Enable Common Custom User Data Maven extension auto-injection` +

> Enables auto-injection of https://github.com/gradle/common-custom-user-data-maven-extension[Common Custom User Data Maven Extension].

Reviewer: ```suggestion
Enables auto-injection of https://github.com/gradle/common-custom-user-data-maven-extension[Common Custom User Data Maven extension].
```

== Gradle Enterprise integration

The plugin can be configured to inject the https://docs.gradle.com/enterprise/gradle-plugin/[Gradle Enterprise Gradle plugin] or the https://docs.gradle.com/enterprise/maven-extension/[Gradle Enterprise Maven extension] into any Gradle or Maven build that is executed on the Jenkins server or on any of its connected agents.
To achieve this, the plugin installs various files on each agent depending on the configuration via the global options in the `Manage Jenkins/Configure System` section.

The same auto-injection behavior is available for the https://github.com/gradle/common-custom-user-data-gradle-plugin[Common Custom User Data Gradle plugin] and the https://github.com/gradle/common-custom-user-data-maven-extension[Common Custom User Data Maven extension].

> **Note**
> - The configuration applies to all builds on all connected agents matching the specified label criteria, or all in case no label criteria are defined.

> == Disabling/uninstalling Gradle Jenkins plugin

Author: Done
